Add file uploading support

MaxLeiter opened the pull request:
Seems to be pretty ready to go.

UI can be improved in another PR.

xPaw commented on defaults/config.js:
```diff
@@ -86,6 +86,28 @@ module.exports = {
+	// @type     int
+	// @default  1e7
+	//
+	uploadMaxFileSize: 1e7,
```
10mb would be 10240.

1e7 is 10 gigabytes.

MaxLeiter replied:
its in bytes not KB

xPaw replied:
Your comment says kilobytes.

The other max size we have in the config is also KB, so do the math in the code.

xPaw commented on src/server.js:
```diff
@@ -281,6 +284,11 @@ function initializeClient(socket, client, token, lastMessage) {
 	client.clientAttach(socket.id, token);
 
+	if (Helper.config.upload) {
+		const upload = require("./plugins/upload");
```
require it once at the top.

xPaw commented on src/plugins/upload.js:
```diff
@@ -0,0 +1,35 @@
+}
+
+/** https://stackoverflow.com/a/12900504 **/
+function getExtension(path) {
```
Use `path.extname`.

xPaw commented on client/js/upload.js:
```diff
@@ -0,0 +1,53 @@
+	return (("draggable" in div) || ("ondragstart" in div && "ondrag" in div)) && "FormData" in window && "FileReader" in window;
+}
+
+if (isDragDropSupported()) {
```
This check isn't really necessary because you're only binding events in the `if`, and they simply won't do anything if there's no support.

Also the fact that all the evergreen browsers have support for it.

xPaw commented on client/js/upload.js:
```diff
@@ -0,0 +1,53 @@
+
+	$form.on("drag dragstart dragend dragover dragenter dragleave", function(e) {
+		e.preventDefault();
+		e.stopPropagation();
```
These 2 functions are shorthanded by `return false` in jQuery.

xPaw commented on client/js/upload.js:
```diff
@@ -0,0 +1,53 @@
+	// Reset progressbar
+	$("#upload-progressbar").width(0);
+	$("#loading-page-message, #connection-error").addClass("shown");
+	$("#loading-page-message, #connection-error").text(event.message);
```
Chain `text` after `addClass`. You also don't need to target `#loading-page-message`.

MaxLeiter replied:
I know I don't need to, but its targeted elsewhere so did this for consistency

xPaw replied:
I know what you mean, but in other places it makes sense because the error can be shown before the app fully loads.

In this case uploads only work after the app loaded, so there is no loading page.

xPaw commented on src/server.js:
```diff
@@ -46,6 +47,9 @@ module.exports = function() {
 		.use("/storage/", express.static(Helper.getStoragePath(), {
 			redirect: false,
 			maxAge: 86400 * 1000,
+		})).use("/uploads/", express.static(Helper.getFileUploadPath(), {
```
We'll want to figure something to make most files have `plain/text` content type, because we don't want to make js/css/etc files arrive with their respective content types.

dgw replied:
Any thoughts on sending `Content-Disposition: inline` as well to encourage browsers to display instead of prompting to save? (If The Lounge saves the original filename anywhere, that could be passed along in this header too, I think.)

MaxLeiter commented on client/index.html.tpl:
```diff
@@ -76,12 +76,18 @@
 						<div class="input">
 							<span id="nick"></span>
 							<textarea id="input" class="mousetrap"></textarea>
+							<span id="submit-tooltip" class="tooltipped tooltipped-w tooltipped-no-touch" aria-label="Upload file">
```
Done

christer88 commented:
https://gfycat.com/ZanyUnsteadyAndalusianhorse was linked earlier on IRC. Shows the lovely uploading feature!

dgw commented:
This excites me so much, even though I probably wouldn't use the feature as-is because of the dependency on The Lounge and its instance's address for serving the images.

The future may bring writing to a custom uploads directory (on a network mount, for example) and returning a custom URL based on that, so uploads can go somewhere besides The Lounge's server if wanted/needed. Maybe it will even support a few third-party hosts (Imgur, etc.) and/or standalone self-hosted image/file servers (Pomf, ownCloud, etc.) via plugins, who knows.

My point is, much of the truly heavy lifting is already in this PR, and I'm positively tingling with excitement for the future of this function. 🎉👏 @MaxLeiter! (Thanks to @christer88 for passing along the video clip of this in action, too!)

Man, it's a great time to be an IRC user. My former IRC friends who've jumped on the Discord bandwagon don't know the Renaissance they're missing! (Even if it is services like Discord that are pushing IRC clients to step up… 😉)

Zarthus commented on defaults/config.js:
```diff
@@ -86,6 +86,28 @@ module.exports = {
+	// Images will not expire or be removed by TheLounge, so be wary of memory and your hardware.
+	//
+	// @type     boolean
+	// @default  false
```
wrong default.

MaxLeiter replied:
No -- wrong value imo. I think the default should be false, as this is unwise on public instances at the moment.

Zarthus replied:
Agreed :)

MaxLeiter replied:
fixed

astorije replied:
I think it's fine to start with until it gets stable, but I think in the long run we should set this to `true` by default:

- Config sets instance in private mode by default
- For discoverability, I think we should have features enabled by default with a way to disable if necessary, but not the other way around (I have the same opinion about prefetch)

McInkay replied:
I mean, discoverability will be better when we get #1950 in and we can add the server settings to the frontend for admins.

Zarthus commented on client/js/socket-events/configuration.js:
```diff
@@ -30,6 +30,10 @@ socket.on("configuration", function(data) {
 	});
 
+	if (!data.upload) {
+		$("#upload").hide();
```
What about making it hidden by default and only showing it if it's enabled? (in event of js error it doesn't look strange then)

xPaw replied:
The UI is not visible at this stage.

MaxLeiter replied:
It's in the index.html, so it will be rendered by default regardless. The alternatives would be to use an underscore template, but that exposes the config option of whether or not it's enabled or not.

MaxLeiter commented on src/server.js:
```diff
@@ -48,6 +50,50 @@ module.exports = function() {
+
+		if (!uploadedFile) {
+			return res.status(404).send("Not found");
+		}
```
Should move this check above setting header so

SinZ163 commented:
What happens when a filename conflict happens?

MaxLeiter commented:
The library automatically renames the file with a number (-0, -1, etc)

xPaw commented on src/plugins/upload.js:
```diff
@@ -0,0 +1,44 @@
+		});
+	};
+
+	siofu.on("complete", (data) => {
```
Handle `error` event (and possibly others? need to look at the source code) so we don't crash.

MaxLeiter replied:
Done

xPaw commented on src/server.js:
```diff
@@ -48,6 +50,50 @@ module.exports = function() {
+		case "video/mp4":
+			break;
+		default:
+			res.set("content-type", "text/plain");
```
If i upload an `exe`, will it get printed in the browser?

MaxLeiter replied:
Now will be served as an attachment

xPaw commented on src/server.js:
```diff
@@ -48,6 +50,50 @@ module.exports = function() {
 			maxAge: 86400 * 1000,
 		}));
 
+	app.get("/uploads/:file", (req, res) => {
```
Perhaps we should move this handler to `uploads.js`.

MaxLeiter replied:
package and theme serving also lives here, fwiw.

xPaw replied:
I know, but we should move it out at some point.

xPaw commented on src/server.js:
```diff
@@ -48,6 +50,50 @@ module.exports = function() {
+		const uploadPath = Helper.getFileUploadPath();
+		const contentType = mime.contentType(path.extname(fileName));
+		const filePath = path.join(uploadPath, fileName);
+		const uploadedFile = upload.ensureFileExists(filePath);
```
Just use `fs.existsSync` directly

xPaw commented on src/plugins/upload.js:
```diff
@@ -0,0 +1,44 @@
+	siofu.on("complete", (data) => {
+		const base = data.file.base;
+		const extension = path.extname(data.file.pathName);
+		const url = `/uploads/${base}${extension}`;
```
Remove leading slash.

MaxLeiter replied:
Done

xPaw commented on src/plugins/upload.js:
```diff
@@ -0,0 +1,44 @@
+const fsextra = require("fs-extra");
+const fs = require("fs");
+
+function init(socket, client) {
```
Convert this file to be a `class`

MaxLeiter replied:
Done

xPaw commented on src/plugins/upload.js:
```diff
@@ -0,0 +1,44 @@
+	siofu.listen(socket);
+}
+
+function ensureFileExists(filePath) {
```
`return fs.existsSync(filePath);`

MaxLeiter replied:
Obsolete now

MaxLeiter commented on src/server.js:
```diff
@@ -48,6 +49,25 @@ module.exports = function() {
 			maxAge: 86400 * 1000,
 		}));
 
+	app.get("/uploads/:name/:slug*?", (req, res) => {
```
https://stackoverflow.com/questions/6161567/express-js-wildcard-routing-to-cover-everything-under-and-including-a-path

xPaw commented on client/js/socket-events/uploads.js:
```diff
@@ -0,0 +1,15 @@
+const wrapCursor = require("undate").wrapCursor;
+
+socket.on("upload:success", (url) => {
+	const fullURL = new URL(url, location);
```
Let's future proof this and check if `url` starts with `http`. If it does then it's a full url.

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,89 @@
+			}
+
+			log.warning(`Failed to read ${filePath}`);
+			return;
```
Change these 2 returns to `null` to match `fileType`.

MaxLeiter replied:
Done

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,89 @@
+		// Returns new file name
+		function handleSaving(data) {
+			const tempPath = path.join(Helper.getFileUploadPath(), ".tmp", data.file.name);
+			const randomName = crypto.randomBytes(16).toString("hex");
```
I think we can get away with shorter random strings. 8 bytes perhaps?

MaxLeiter replied:
Done

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,89 @@
+		function handleSaving(data) {
+			const tempPath = path.join(Helper.getFileUploadPath(), ".tmp", data.file.name);
+			const randomName = crypto.randomBytes(16).toString("hex");
+			const destPath = path.join(Helper.getFileUploadPath(), randomName);
```
Take first 2 bytes of the string and make it a folder so that it's split into subfolders like preview storage does it.

However keep just the hash in URL (no need to show subfolder to the user).

MaxLeiter replied:
Done

xPaw replied:
You haven't done this.

xPaw replied:
Nevermind, you actually have. My bad!

xPaw commented on src/server.js:
```diff
@@ -48,6 +49,25 @@ module.exports = function() {
 		}));
 
+	app.get("/uploads/:name/:slug*?", (req, res) => {
+		const name = req.params.name;
```
Add a length (and perhaps a-z0-9?) check to match the hash format we generate. This will avoid trying to request `.tmp` folder as well.

MaxLeiter replied:
Im a little hesitant to enforce too strict rules regarding this because plugins may want to do something differently and there’s no easy way for them to interface with express. But that’s a ways off, so I’m fine with it for now

xPaw replied:
That's one of the reasons I wanted you to move this route to the upload plugin, because I expect other plugins to provide their own route if they need it. I expect other plugins to provide custom URLs though (imgur, etc).

MaxLeiter replied:
Is `[0-9a-fA-F]+` fine with you?

Need to figure out best way to access express in `uploader.js` in order to handle routing there

MaxLeiter replied:
Fixed with what we discussed on IRC

xPaw replied:
For the route, you can do `Uploader.router(app);`

MaxLeiter commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,89 @@
+
+			// Try again
+			if (fs.exists(destPath)) {
+				return handleSaving(data);
```
should be `while (!fs.exists(destPath))`, right?

xPaw replied:
No, you want it to keep trying to find a filename if one it generated exists already.

MaxLeiter replied:
```
path.js:57
function normalizeString(path, allowAboveRoot, separator, isPathSeparator) {
                        ^

RangeError: Maximum call stack size exceeded
    at normalizeString (path.js:57:25)
    at Object.normalize (path.js:1133:12)
    at Object.join (path.js:1168:18)
    at handleSaving (/Users/maxleiter/Documents/lounge/src/plugins/uploader.js:51:26)
```

MaxLeiter replied:
Fixed

xPaw commented on src/server.js:
```diff
@@ -48,9 +49,41 @@ module.exports = function() {
+
+		// If whitelisted
+		if (fileType && fileType.mime && Uploader.isValidType(fileType.mime)) {
+			mimeType = fileType.mime;
```
I think `mimeType` should be set regardless. And `Uploader.isValidType` will control `contentDisposition`.

Are we going to look into using `is-utf8` as we discussed?

MaxLeiter commented:
Do you mean move most of this code to getFileType?

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,121 @@
+		siofu.on("complete", (data) => {
+			const randomFileName = handleSaving(data);
+
+			if (!randomFileName) {
```
This `if` doesn't appear to be necessary.

Jay2k1 commented:
Referencing #1679 which is a feature request for this

xPaw commented on client/js/socket-events/init.js:
```diff
@@ -29,6 +29,7 @@ socket.on("init", function(data) {
 
 	$("#connection-error").removeClass("shown");
 	$(".show-more button, #input").prop("disabled", false);
+	$("#upload").show();
```
Is this gonna show itself up upon reconnect if uploads are disabled?

xPaw commented on client/js/socket-events/uploads.js:
```diff
@@ -0,0 +1,15 @@
+	wrapCursor(textbox, fullURL, "");
+});
+
+socket.on("upload:failure", () => {
```
This is never emitted.

MaxLeiter replied:
Leave it in for plugins?

xPaw replied:
Let's worry about that later, remove it for now.

MaxLeiter replied:
done

MaxLeiter commented:
Plugins may choose to emit it?

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,156 @@
+				destPath = path.join(Helper.getFileUploadPath(), randomName.substring(0, 2), randomName);
+			} while (fs.exists(destPath));
+
+			fsextra.moveSync(tempPath, destPath, (err) => {
```
Let's use the async version. It should be really simple as adding `callback` and calling it with `randomName` in this move callback.

xPaw commented on client/index.html.tpl:
```diff
@@ -76,12 +76,18 @@
 					<script async src="js/loading-error-handlers.js"></script>
 				</div>
 				<div id="chat-container" class="window">
+					<div id="chat-container-overlay"></div>
```
Move this overlay to body, next to the image viewer. Keeping it here means you can't start dragging over the channel list.

And give this a better id like `upload-overlay` or something.

MaxLeiter replied:
I think you can start dragging over channel lost, as long as you drag on to chat container after, right? The onDrop listener is only on the chat container, not the channel list as well

xPaw replied:
Yeah  change the onDrop to be on the entire document. It's rather weird UX to drag a file into Lounge and have it opened in that tab instead of being uploaded.

MaxLeiter replied:
done

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,156 @@
+	}
+
+	static isValidType(mimeType) {
+		const whitelist = [
```
Move this whitelist to the top (after requres)

MaxLeiter replied:
Fixed

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,156 @@
+				return;
+			}
+
+			log.warning(`Failed to read ${filePath}`);
```
It's log.warn not log.warning.

MaxLeiter replied:
Fixed

xPaw commented:
Some things to test/do before shipping this feature:

- How is the upload handled when you try to drag multiple files onto the window
 - What about selecting multiple files in the file select?
- What happens when you try to drag a file (link) from another browser tab
- Should we have a confirmation dialog before doing the actual upload? We can also render a small thumbnail using browser technology.

xPaw commented on client/js/socket-events/init.js:
```diff
@@ -31,6 +31,10 @@ socket.on("init", function(data) {
 	$("#submit").show();
 
+	if (data.uploads) {
+		$("#upload").show();
```
As discussed in IRC, let's keep hide/show in configuration event.

MaxLeiter replied:
Done

xPaw commented on client/js/socket.js:
```diff
@@ -11,6 +11,12 @@ const socket = io({
 });
 
+$("#loading-page-message, #connection-error").on("click", function() {
+	if ($(this).hasClass("shown")) {
```
Unnecessary check.

xPaw commented on client/js/upload.js:
```diff
@@ -0,0 +1,42 @@
+	$("#connection-error").addClass("shown").text(event.message);
+});
+
+const $form = $("#chat-container");
```
I think this target should be `$(document)`.

MaxLeiter replied:
Done

xPaw commented on defaults/config.js:
```diff
@@ -86,6 +86,28 @@ module.exports = {
+	// Maximum file size for uploaded files
+	//
+	// If upload is enabled, The Lounge will only allow uploading files with a size < the value below
+	// Specified value is in kilobytes. Default value is 10240 bytes (10 MB).
```
Default value is *kilo*bytes.

Also probably should be 10000 since it gets multiplied by 1024? (I know I said 10240 myself before).

MaxLeiter replied:
As other config values are specific (prefetch is 2048), i'll instead multiply by 1000 for consistency.

MaxLeiter replied:
Done (my comment)

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,154 @@
+			const folder = path.join(Helper.getFileUploadPath(), ".tmp");
+			fsextra.ensureDir(folder, (err) => {
+				if (err) {
+					callback(err);
```
Any reason this folder creator is in `uploadValidator` and not simply in the constructor?

And looking at the library code, err shouldn't be passed through to correctly abort the error: 
https://github.com/sffc/socketio-file-upload/blob/c1946fe3e43bd0692100c350c93771ec01ac0823/server.js#L314-L316

MaxLeiter replied:
No, except it exists in the library for basically this purpose. I can do either way

xPaw replied:
I think we will be fine just creating the folder in the constructor.

Instead of error callback, print a warning to the console.

MaxLeiter replied:
if you want a warning in console on error, do you mean I should leave the `ensureDir` in the uploadValidator? If we're creating the directory in the constructor, then no need to override uploadValidator.

MaxLeiter replied:
Nevermind, got it

MaxLeiter replied:
done

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,154 @@
+		siofu.on("complete", (data) => {
+			handleSaving(data).then((randomName) => {
+				const randomFileName = randomName;
+				const slug = data.file.name;
```
I think `file.base` should be the cleaned up and safer name, so we probably should use that.

MaxLeiter replied:
We don't want file extensions?

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	static isValidType(mimeType) {
+		return whitelist.indexOf(mimeType) > -1;
```
We can do `return whitelist.includes(mimeType);` here.

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,154 @@
+			const filePath = path.join(uploadPath, folder, name);
+			const nameRegex = /^[0-9a-f]{16}$/;
+
+			if (!nameRegex.test(name)) {
```
Move this `if` higher, right after defining `name` const.

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,154 @@
+	static router(express) {
+		express.get("/uploads/:name/:slug*?", (req, res) => {
+			const name = req.params.name;
+			const folder = req.params.name.substring(0, 2);
```
`req.params` is unnecessary here as you have it in `name` variable.

MaxLeiter commented on client/js/upload.js:
```diff
@@ -0,0 +1,42 @@
+const $form = $("#chat-container");
+const $overlay = $("#upload-overlay");
+
+$form.on("drag dragstart dragend dragover dragenter dragleave", () => false);
```
Done

MaxLeiter commented:
xPaw: Followed your comments, along the way the overlay broke, and I cant for the life of me figure it out. Will look into more once other comments addressed.

astorije commented:
> Should we have a confirmation dialog before doing the actual upload? We can also render a small thumbnail using browser technology.

Not sure what you mean by browser technology.
I'm not sure a dialog is necessary, but I think it would be nice to have (oh god how am I going to explain this clearly) thumbnail(s) of what just got uploaded above the message input, each with a &times; button in the corner. When clicked, the thumbnail disappears, and on the backend we destroy the upload.

Either way, I think this is waaay beyond your PR :) I'd be absolutely fine with what you currently have: no confirmation, no UI, etc. Just the bare feature, and then we build up on it!

MaxLeiter commented:
I think xPaw meant we can show the file before its uploaded in the dialog client-side (so we can display the preview without it being uploaded).

I'm all for no confirmation UI this PR :)

xPaw commented:
> I think, for the initial version, we may want to display an error (using disconnect GUI) saying multiple file uploads is not yet supported?

Fine by me.

xPaw commented on client/js/upload.js:
```diff
@@ -0,0 +1,42 @@
+const $form = $(document);
+const $overlay = $("#upload-overlay");
+
+$form.on("drag dragstart dragend dragleave", () => false);
```
Remove `dragleave` here.

And add `return false` in the next 2 events, I guess that's why you had them multiplied.

MaxLeiter replied:
Done

xPaw commented on client/js/upload.js:
```diff
@@ -0,0 +1,42 @@
+
+$form.on("drag dragstart dragend dragleave", () => false);
+
+$form.on("dragover dragenter", () => {
```
Do we actually need `dragover` and `dragleave` events?

Looking at MDN tells me we kind of don't.

https://developer.mozilla.org/en-US/docs/Web/Events/dragenter
https://developer.mozilla.org/en-US/docs/Web/Events/dragover
https://developer.mozilla.org/en-US/docs/Web/Events/dragend
https://developer.mozilla.org/en-US/docs/Web/Events/dragleave

MaxLeiter replied:
Nope

astorije commented:
> The idea is to review the file you have dragged in before being automatically uploaded. I have tried at least 3 different chats, and they all asked.
> 
> But I guess the fact we put an URL in input instead of sending it also helps here.

Both statements make sense to me, and I agree. I think we're all on the same page, and also that it should go in a future PR, so I'm not going to spend too much time thinking about UI/UX for this just now. But thanks for clarifying!

astorije commented on client/css/style.css:
```diff
@@ -296,6 +297,7 @@ kbd {
 #footer .settings::before { content: "\f013"; /* http://fontawesome.io/icon/cog/ */ }
 #footer .help::before { content: "\f059"; /* http://fontawesome.io/icon/question/ */ }
 
+#form #upload::before { content: "\f093"; /* http://fontawesome.io/icon/upload/ */ }
```
Super minor: URL is https://fontawesome.com/icons/upload?style=solid

However, I'm not a huge fan of this icon, not very pretty. I went through the ones FA provide and I think the ones below look better (more or less ordered by preference):

- https://fontawesome.com/icons/paperclip?style=solid
- https://fontawesome.com/icons/cloud-upload-alt?style=solid
- https://fontawesome.com/icons/file-image?style=solid
- https://fontawesome.com/icons/share-square?style=solid
- https://fontawesome.com/icons/desktop?style=solid or https://fontawesome.com/icons/laptop?style=solid (maybe far fetched)

What do you think?

xPaw replied:
`paperclip` makes sense

MaxLeiter replied:
Done

astorije commented:
@MaxLeiter, I'm getting the following when I test your PR with `./scripts/run_pr.sh 2552`:

```
.../thelounge/src/server.js:617
	config.uploadMaxFileSize = uploader.getMaxFileSize();
	                                    ^

TypeError: Cannot read property 'getMaxFileSize' of null
    at getClientConfiguration (.../thelounge/src/server.js:617:38)
    at initClient (.../thelounge/src/server.js:638:32)
    at authCallback (.../thelounge/src/server.js:688:3)
    at Socket.performAuthentication (.../thelounge/src/server.js:702:11)
```

astorije commented on src/server.js:
```diff
@@ -603,6 +614,8 @@ function getClientConfiguration(network) {
 		config.defaults.nick = Helper.getDefaultNick();
 	}
 
+	config.uploadMaxFileSize = uploader.getMaxFileSize();
```
If I read this correctly, when `Helper.config.upload` is `false` (or not set in the config, i.e. coming from defaults which are `false`), `uploader` hasn't been set so it's `undefined` 💥

MaxLeiter commented:
@astorije: fixed

@astorije, @xPaw:
Moved upload code to an `init()` function called in the configuration socket event. This way the file upload overlay won't show and SIOFU wont be instantiated if its not enabled on the server.

xPaw commented on client/js/socket-events/configuration.js:
```diff
@@ -30,6 +31,14 @@ socket.on("configuration", function(data) {
 	});
 
+	if (data.upload) {
+		$("#upload").show();
```
Just realized that this show/hide needs to be moved before `if (options.initialized)`.

Keep the init here though.

MaxLeiter replied:
done

xPaw commented on client/js/socket.js:
```diff
@@ -11,6 +11,10 @@ const socket = io({
 	reconnection: !$(document.body).hasClass("public"),
 });
 
+$("#loading-page-message, #connection-error").on("click", function() {
```
Remove `#loading-page-message` here, we don't want it clickable.

MaxLeiter replied:
done

MaxLeiter commented on client/js/upload.js:
```diff
@@ -0,0 +1,54 @@
+const SocketIOFileUpload = require("socketio-file-upload/client");
+const instance = new SocketIOFileUpload(socket);
+
+function init() {
```
done

xPaw commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,157 @@
+	"audio/webm",
+	"video/webm",
+	"audio/ogg",
+	"video/ogg ",
```
Hey there's a space here

MaxLeiter replied:
I did this on purpose to differentiate. I can remove if youd prefer

xPaw replied:
Differentiate what? Having a space in the string means this type won't be found.

MaxLeiter commented:
Relies on #2592

McInkay commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,146 @@
+const isUtf8 = require("is-utf8");
+const log = require("../log");
+
+const whitelist = [
```
The plan is to open up this whitelist later on, or allow it to be turned off, yeah?

MaxLeiter replied:
Up for discussion. It's tied to the plugin, so future plugins can use their own.

McInkay replied:
Yeah, I wasn't meaning for right now, but it can be a discussion later. Cool beans.

McInkay commented on defaults/config.js:
```diff
@@ -86,6 +86,28 @@ module.exports = {
+	// @type     int
+	// @default  10240
+	//
+	uploadMaxFileSize: 10240,
```
Can we allow this to be set to `-1` to be disabled?

MaxLeiter replied:
Previews dont have a max size, but sure.

McInkay replied:
What do you mean?

MaxLeiter replied:
Previews have a max fetch size, but we have no way of setting unlimited. So this just isnt very consistent

McInkay replied:
Ah right, well we can leave it for just now if we don't do it anywhere else.

McInkay commented on package.json:
```diff
@@ -44,19 +44,23 @@
     "cheerio": "0.22.0",
     "commander": "2.16.0",
     "express": "4.16.3",
+    "file-type": "8.0.0",
```
Just before you merge, it's probably worth checking these are up to date, as I'm pretty sure we only get PRs for updates after it's merged in.

MaxLeiter replied:
Yep, updated some. Good call.

McInkay commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,146 @@
+
+class Uploader {
+	constructor(client, socket) {
+		const siofu = new SocketIOFileUploadServer();
```
Can we rename this? It's pretty confusing.

MaxLeiter replied:
I disagree, it's pretty self-explanatory. It's what the SIOFU docs use, so I think thats a fine reason for keeping as-is. I don't care too much if you feel strongly.

McInkay replied:
It's self explanatory after you go up and read the name and go "oh, I guess that's what it means" but it's really quick to forget what that means. But if you read it down you have 0 idea, no clue whatsoever what it is.

MaxLeiter replied:
My argument is I think anyone editing the file should have an idea of what it is, and this way we stay consistent with the docs

McInkay replied:
I disagree.

I think that the "this should be slightly confusing so that you know what you are doing" is a bad trope that exists in software development that I think we should be getting rid of. It's an elitist view, imo, which is really not great for software development as a whole. We should be making things easier for people, not more difficult.

And there's no need to stay consistent with the docs, the docs have named something in a confusing way, let's get rid of that.

MaxLeiter replied:
Fixed

MaxLeiter commented on defaults/config.js:
```diff
@@ -139,6 +139,29 @@ module.exports = {
 	// This value is set to `2048` kilobytes by default.
 	prefetchMaxImageSize: 2048,
 
+	// ### `upload`
```
@astorije: if you could review this file that'd be great

astorije replied:
Nitpick: Do we want to call this `fileUpload`? `upload` alone could be confusing, especially as we add new features in the future. 🤷‍♂️

astorije replied:
So, an alternative to:

```
fileUpload: true,
fileUploadMaxSize: 10240,
```

is to nest it like we do in a few other places:

```
fileUpload: {
  enabled: true,
  maxFileSize: 10240,
}
```

Thoughts?

xPaw commented on package.json:
```diff
@@ -44,19 +44,23 @@
     "cheerio": "0.22.0",
     "commander": "2.17.1",
     "express": "4.16.3",
+    "file-type": "8.1.0",
```
Update to last version

xPaw commented on package.json:
```diff
@@ -44,19 +44,23 @@
     "request": "2.88.0",
     "semver": "5.5.0",
     "socket.io": "2.1.1",
+    "socketio-file-upload": "0.6.1",
```
Update to last version

MaxLeiter replied:
done

astorije commented on defaults/config.js:
```diff
@@ -139,6 +139,29 @@ module.exports = {
+	// Allow uploading files to the server hosting TheLounge.
+	// If enabled, The Lounge will store files in the `${THELOUNGE_HOME}/uploads` folder.
+	//
+	// Images will not expire or be removed by TheLounge, so be wary of memory and your hardware.
```
Typo: `TheLounge` (and at L144).

Instead of `so be wary of memory and your hardware`, I would go with `which can cause issues depending on your hardware, for example in terms of disk usage.` No big deal though.

astorije commented on defaults/config.js:
```diff
@@ -139,6 +139,29 @@ module.exports = {
+	//
+	// Maximum file size for uploaded files
+	//
+	// If upload is enabled, The Lounge will only allow uploading files with a size < the value below
```
This seems redundant with the line above. Instead, I would phrase it to explain what happens when a file above this limit is sent. Something like:
`When file upload is enabled, users sending files above this limit will be prompted an error message in their browser.` (I have no clue if this is remotely accurate as I have not reviewed the rest of the PR yet, it's just to illustrate).

astorije commented on defaults/config.js:
```diff
@@ -139,6 +139,29 @@ module.exports = {
+	//
+	// `-1` will disable the file size limit and will allow files of any size. Use at your own risk.
+	//
+	// This value is set to `10240` by default.
```
You are specifying this twice.

astorije commented on defaults/config.js:
```diff
@@ -139,6 +139,29 @@ module.exports = {
+	// `-1` will disable the file size limit and will allow files of any size. Use at your own risk.
+	//
+	// This value is set to `10240` by default.
+	//
```
Remove this empty line

xPaw commented:
@astorije I fixed your comments.

astorije commented on client/js/socket.js:
```diff
@@ -11,6 +11,10 @@ const socket = io({
 
+$("#connection-error").on("click", function() {
+	$(this).removeClass("shown");
+});
```
I'm not sure to understand what this is for. My understanding is that the banner is now clickable and clicking it removes it. What's the reason behind it?

xPaw replied:
Upload errors re-use that red banner.

astorije commented on defaults/config.js:
```diff
@@ -139,6 +139,27 @@ module.exports = {
+	// `-1` will disable the file size limit and will allow files of any size. Use at your own risk.
+	fileUpload: {
+		enable: true,
+		maxFileSize: 10000,
```
@xPaw, why did you change this from 10240 to 10000? Other fields (well, at least `prefetchMaxImageSize`) are a power of 2, and it's also more predictable for users as most OS show sizes in MB. I'd rather remember to check that a file is < 10MB and not 9.77MB :)

xPaw replied:
I changed the multiplier to 1024 to match the other max size field we have in the config already.

astorije replied:
Right, but you made the change so that it multiplies 10000kb into 10000*1024 bytes, this value should still be 10240 so it's 10mb (10*1024*1024 = 10,485,760 bytes)... no?
At the very least, `10240` here would be consistent with the `2048` above.

xPaw replied:
You're right, fixed.

astorije commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,152 @@
+	"image/jpeg",
+	"image/png",
+	"image/webp",
+	"text/plain",
```
Just curious, things like log files (often in `.log`) and Markdown files (`.md`), they would be  treated as plain text, right? (I haven't tested yet, doing code review first then I'll play with the feature)

xPaw replied:
The extension doesn't matter, it detects the type from the file contents, so yes logs or markdown will work fine.

astorije replied:
Awesome, that's what I thought, thanks for confirming!

astorije commented on src/plugins/uploader.js:
```diff
@@ -0,0 +1,152 @@
+
+		if (configOption === -1) { // no file size limit
+			return null;
+		}
```
Do we have a protection against negative values (or zero) somewhere? Not sure how the uploader library reacts on 0 or -42, maybe we should blow up the process ourselves?

xPaw replied:
Looked at the source code, it will simply act as a limit (say file exceeded limit)

`if (self.maxFileSize !== null && file.size > self.maxFileSize) {`

astorije commented on defaults/config.js:
```diff
@@ -139,6 +139,27 @@ module.exports = {
+	//
+	// The available keys for the `fileUpload` object are:
+	//
+	// - `enable`: When `true`, the identd daemon runs on server start.
```
@xPaw, I made the following fixes/changes here:

<details>

```diff
@ config.js:145 @ module.exports = {
    // ### `fileUpload`
    //
    // Allow uploading files to the server hosting The Lounge.
-   // If enabled, The Lounge will store files in the `${THELOUNGE_HOME}/uploads` folder.
    //
-   // Files will not expire or be removed by The Lounge, which can cause issues
-   // depending on your hardware, for example in terms of disk usage.
+   // Files are stored in the `${THELOUNGE_HOME}/uploads` folder, do not expire,
+   // and are not removed by The Lounge. This may cause issues depending on your
+   // hardware, for example in terms of disk usage.
    //
    // The available keys for the `fileUpload` object are:
    //
-   // - `enable`: When `true`, the identd daemon runs on server start.
-   // - `maxFileSize`: When file upload is enabled, users sending files above this
-   //    limit will be prompted an error message in their browser.
-   //    Specified value is in kilobytes. Default value is 10000 kilobytes (10 MiB).
-   //
-   // `-1` will disable the file size limit and will allow files of any size. Use at your own risk.
+   // - `enable`: When set to `true`, files can be uploaded on the client with a
+   //   drag-and-drop or using the upload dialog.
+   // - `maxFileSize`: When file upload is enabled, users sending files above
+   //   this limit will be prompted an error message in their browser. A value of
+   //   `-1` disables the file size limit and allows files of any size. **Use at
+   //   your own risk.** This value is set to `10000` kilobytes by default.
    fileUpload: {
        enable: true,
        maxFileSize: 10000,
```
</details>

This is how it looks in the docs page:

<img width="549" alt="screen shot 2018-09-03 at 00 15 46" src="https://user-images.githubusercontent.com/113730/44967723-53cc4180-af11-11e8-822a-4f5a179eedb0.png">

Let me know what you think :)

MaxLeiter commented:
🎉

mostlyvirtual commented:
Thank you for implementing this. What was the reasoning of having this disabled by default? Also, would it be possible to include the enable/disable toggle in the UI settings?
